### ISSUE TITLE
Support Python/C++ interoperability (native Python modules)

### ISSUE THREAD

**Issue by davidzchen** [bazelbuild/bazel#701]
According to the [documentation for `py_library.deps`](http://bazel.io/docs/be/python.html#py_library.deps):

> The list of other libraries to be linked in to the binary target. See general comments about deps at Attributes common to all build rules. These can be py_binary rules, py_library rules or cc_library rules,

However, when I try to build a `py_library` target that depends on a `cc_library`, I get the following error:

```
❯❯❯ bazel build //external:markupsafe
ERROR: /private/var/tmp/_bazel_dzc/a4c09a1fcefb8b21c5b0e7d7db6f08ba/external/markupsafe_archive/BUILD:9:12: in deps attribute of py_library rule @markupsafe_archive//:markupsafe: cc_library rule '@markupsafe_archive//:speedups' is misplaced here (expected py_binary or py_library).
ERROR: Analysis of target '//external:markupsafe' failed; build aborted.
INFO: Elapsed time: 0.213s
```

According to [`BazelPyRuleClasses.java`](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/bazel/rules/python/BazelPyRuleClasses.java), the only allowed rule types in `py_library.deps` is, in fact, `py_binary` and `py_library`:

``` java
  public static final String[] ALLOWED_RULES_IN_DEPS = new String[] {
      "py_binary",
      "py_library",
  };
```

What is the plan for Python-C interoperability? Should we update the documentation to remove the mention of depending on `cc_library` in the meantime?

The background for this is that I am trying to use [Jinja](http://jinja.pocoo.org/) for generating HTML for the Skylark docgen tool that I am working on. However, Jinja depends on [Markupsafe](http://www.pocoo.org/projects/markupsafe/), which is partly implemented in C.

**Comment by davidzchen** [comment]
Some related discussion here: https://groups.google.com/forum/#!searchin/bazel-discuss/pex/bazel-discuss/-DFOrVqK8aY/-7h4e69hfkEJ

**Comment by ulfjack** [comment]
Recategorizing this as a feature request to support Python-C++ interop. I marked the other report that asked for that as a dupe.

**Comment by evanj** [comment]
A workarounds I found: you can include pre-compiled `.so`s in the `py_library`'s `data` attribute. No idea if you can somehow get the output of a `cc_library` to work in this way.

**Comment by benley** [comment]
The trick to building a .so that works as a python extension seems to be making it a `cc_binary` instead of a `cc_library`.  I think you have to build it with `linkshared = 1` to make it work, but I'd love to be wrong about that.

**Comment by MarkusTeufelberger** [comment]
I stumbled upon this one also recently. Please at least change the documentation or (ideally) increase the priority to get this fixed for good. Python libraries and binaries can and do depend on C libraries. This is currently not really easy to model in bazel.

In the meantime it would also be nice to have this workaround documented somewhere.

**Comment by buchgr** [comment]
@MarkusTeufelberger we love contributions 😃

**Comment by ulfjack** [comment]
The problem is that we don't know how to fix it. We could export our internal model, but we're unsure if that's what people want. Our internal model is that all cc_library rules in deps of py_library rules get merged into a single .so at the py_binary level, together with all cc_library dependencies of py_extension rules, except the py_extension code itself, which gets compiled into a separate .so. It sounds awkward, and it is, but there doesn't seem to be any other principled approach to avoid ODR violations in the C++ code.

**Comment by ThomasColthurst** [comment]
+1 for exporting the internal model.  My project is considering switching from bazel to Makefiles solely to avoid the ODR violations we are currently seeing when using bazel + clif + protobufs.

**Comment by mhlopko** [comment]
Work on this is blocked by https://github.com/bazelbuild/bazel/issues/4570.

**Comment by brandjon** [comment]
See related discussion in #1475.

**Comment by petermelias** [comment]
Is there any update on this / roadmap to look at? Trying to plan ahead as would like to use Bazel for building x-plat Docker images which include the Postgres (psycopg2) driver that relies on the host system having libpq available. Cannot currently do this is in a cross-platform / hermetic fashion as far as I can tell currently.

Host machine = OSX, Docker image = Debian

Work-around suggestions welcome though.

@brandjon

**Comment by quval** [comment]
I have a draft of a Starlark implementation of something like the internal Google model (all native deps are linked together into a statically-linked shared object; the modules themselves are dynamically linked to it). Would it be interesting to share this? Perhaps on rules_python?

**Comment by pvcnt** [comment]
@quval It would certainly be interesting yes. :)

**Comment by quval** [comment]
OK, here goes: https://github.com/quval/rules_native_python/commit/f8eaecd9b124b433ec083fd71e0bef044cb04c6c
Got it to work on both Linux and Mac, but only with a toy setup, so I'd still consider this a draft. More importantly, I'm not sure I like the hack I did to work around the circular dependency (native deps -- compiled into native module -- linked against native deps). Any thoughts and suggestions are welcome.

**Comment by mfarrugi** [comment]
The following implementation seems to work for me without issue, are there any deficiencies I'm overlooking?

```python
# py_native_library.bzl

def py_native_library(name, **kwargs):
	native.cc_binary(
		name = name + ".so",
		linkshared = True,
		**kwargs
	)
	native.py_library(
		name = name,
		data = [name + ".so"],
	)
```

**Comment by kkpattern** [comment]
> The problem is that we don't know how to fix it. We could export our internal model, but we're unsure if that's what people want. Our internal model is that all cc_library rules in deps of py_library rules get merged into a single .so at the py_binary level, together with all cc_library dependencies of py_extension rules, except the py_extension code itself, which gets compiled into a separate .so. It sounds awkward, and it is, but there doesn't seem to be any other principled approach to avoid ODR violations in the C++ code.

Hi ulfjack, how doesn't google handle multiple py_library with some common cc_library depends? For example, if I have two py_library both depends on some X cc_library. Initially we may use those two py_library seperately, but at some point people may use both of these two py_library. I think in this situation there is still ODR violations?

**Comment by vkresch** [comment]
@mfarrugi thanks for your example. This works when I am using pybind11. Meaning if I want to call C++ functions via python this works. But how about the other way around call python functions from C++. Can I do this then:

```python
def cc_native_library(name, **kwargs):
	native.py_library(
		name = name + ".py",
		linkshared = True,
                linkstatic = False,
		**kwargs
	)
	native.cc_library(
		name = name,
		data = [name + ".py"],
	)
```

I tried this approach for pybind11 but the problem is here with the third party packages which do not get imported correctly since they are used in the data argument of cc_library. Any suggestions? Thanks in advance!

**Comment by jab** [comment]
@bazelbuild/triage
